Sources/StreamChat/StateLayer/ChatState+Internal.swift

```diff
@@ -0,0 +1,92 @@
+            return lookup.firstUnreadMessage
+        }
+        
+        private var firstUnreadMessage: MessageId? {
+            guard let readInfo = readStates.first(where: { $0.user.id == userId }) else {
+                // Read state is unavailable
+                return hasLoadedAllPreviousMessages ? oldestRegularMessageId : nil
+            }
+            guard readInfo.unreadMessagesCount > 0 else { return nil }
+            guard let lastReadMessageId = readInfo.lastReadMessageId else {
+                // Everything is unread if read state is there but there is no lastReadMessageId
+                return hasLoadedAllPreviousMessages ? oldestRegularMessageId : nil
+            }
+            if let lastReadIndex = indexOfMessage(lastReadMessageId) {
+                if isMostRecent(at: lastReadIndex) {
+                    // Everything has been read
+                    return nil
+                } else {
+                    return lookupUnreadMessage(after: lastReadIndex)
+                }
+            } else {
+                // Can't reach the last read message (if all have been loaded then the channel might have been truncated or hidden, in that case, use the oldest message)
+                return hasLoadedAllPreviousMessages ? oldestRegularMessageId : nil
+            }
+        }
+        
+        private func lookupUnreadMessage(after excludedSearchIndex: Int) -> MessageId? {
+            let searchRange: ReversedCollection<Range<Int>> = {
+                if messageOrder.isAscending {
+                    return (messages.endIndex..<excludedSearchIndex).reversed()
+                } else {
+                    return (messages.startIndex..<excludedSearchIndex).reversed()
+                }
+            }()
+            for index in searchRange {
+                if let message = messages[safe: index], message.deletedAt == nil, message.author.id != userId {
+                    return message.id
+                }
+            }
+            return nil
+        }
+        
+        private func indexOfMessage(_ messageId: MessageId) -> Int? {
+            messages.firstIndex(where: { $0.id == messageId })
+        }
+        
+        private func isMostRecent(at index: Int) -> Bool {
+            if messageOrder.isAscending {
+                return messages.index(before: messages.endIndex) == index
+            } else {
+                return messages.startIndex == index
+            }
+        }
+        
+        private var oldestRegularMessageId: MessageId? {
+            if messageOrder.isAscending {
+                return messages.first(where: \.isRegular)?.id
+            } else {
+                return messages.last(where: \.isRegular)?.id
+            }
+        }
+    }
+}
+
+private extension ChatMessage {
+    var isRegular: Bool {
```

Comment:
why do we need this?

Reply:
@martinmitrevski @laevandus  I remember this is needed for edge cases. Not sure which edge cases at the moment, but this is 100% needed.

Reply:
Btw, all this logic will go away, since all SDKs have agreed to make this simpler, so it could be an opportunity to do it here 🤔